### PR TITLE
feat(aws-rds): support Aurora Serverless v2 cluster and instances

### PR DESCRIPTION
**Pull request by pahud**
## About this PR 

Aurora Serverless v2 now [supports](https://aws.amazon.com/about-aws/whats-new/2022/10/amazon-aurora-serverless-v2-supports-aws-cloudformation/) AWS CloudFormation and CDK has included the L1 support. The cluster with serverless v2 support is created from ([AWS::RDS::DBCluster](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbcluster.html)) with some conditions and limitations:

- When [ServerlessV2ScalingConfiguration](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbcluster.html#cfn-rds-dbcluster-serverlessv2scalingconfiguration) is enabled, the DBCluster will be created as a cluster with serverless v2 support so we can add serverless v2 writer and reader instances in this cluster.
- Serverless v1 and v2 have different scaling spec.
- Clusters with serverless v2 support  can add either `serverless` instances or `provisioned` instances, which means when `DBCluster` is provisioned with v2 support, we should be able to add an arbitrary number of instances in it, either `serverless` or `provisioned`. The first one added would be the writer while others the readers.

This PR extends the `DatabaseCluster` construct so we can enable its serverless v2 support by specifying the new `serverlessV2Scaling` property. After the cluster is created, we can `addServerlessInstance` to add a serverless v2 instance or `addInstance` to add a provisioned instance into the serverless v2 enabled DBCluster.

This PR has tested a few scenarios in the `integ.serverlessv2-cluster.ts` including:

- DBCluster with 1 master and 0 or more replica instances(we already can satisfy this with the existing DatabaseCluster L2)
- DBCluster with 1 serverless v2 write instance only
- DBCluster with 1 serverless v2 write and 1 or more serverless v2 read instances
- DBCluster with 1 serverless v2 write and 1 or more provisioned read instances 
- DBCluster with 1 provisioned write and 1 or more serverless v2 read instances

## call out

I am not sure if this would make breaking changes:

- `instances` from of `DatabaseClusterBaseProps` can be `0` now. This allows the DBCluster to be provisioned as a cluster with serverless instances only.

Fixes: #20197 

----

### All Submissions:

* [ ] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts [hunk 1]
```diff
@@ -462,6 +512,34 @@ abstract class ServerlessClusterNew extends ServerlessClusterBase {
...
+    const maxCapacity = options.maxCapacity;
+
+    // maxCapacity should be higher than 0.5 and we can't specify 0.5 for both minCapacity and maxCapacity so maxCapacity should be at least 1
+    // @see https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.setting-capacity.html#aurora-serverless-v2.max_capacity_considerations
+    if (minCapacity && minCapacity < 0.5 || maxCapacity && maxCapacity < 1) {
+      throw new Error('the smallest value that you can use for minCapacity and maxCapacity is 0.5 and 1');
+    }
+
+    if (maxCapacity && maxCapacity > 128) {
+      throw new Error('the largest value that you can use for maxCapacity is 128');
```
Comment (moltar) [inline packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts:527]:
```suggestion
      throw new Error('The largest value that you can use for maxCapacity is 128.');
```

File: packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts [hunk 2]
```diff
@@ -121,6 +129,13 @@ interface ServerlessClusterNewProps {
    */
   readonly scaling?: ServerlessScalingOptions;
 
+  /**
+   * Scaling configuration of an Aurora Serverless V2 database cluster.
+   *
+   * @default - TBC
+   */
+  readonly scalingV2?: ServerlessV2ScalingOptions;
```
Comment (moltar) [inline packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts:137]: That `V2` suffix is a bit of an eye sore.

Is there no better way?

Can we perhaps set an enum on the  `ServerlessCluster` class that specifies which version it is, and it then type narrows the interface? Or will it confuse JSII?

```ts
enum ServerlessClusterVersion {
  V1,
  V2,
}

interface ServerlessClusterNewProps {
  version: ServerlessClusterVersion
}

interface ServerlessClusterNewPropsV1 {
  version: ServerlessClusterVersion.V1
  scaling: V1Scaling
}

interface ServerlessClusterNewPropsV2 {
  version: ServerlessClusterVersion.V2
  scaling: V2Scaling
}
```

**Comment by moltar** [comment]
Also, from the docs: https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.how-it-works.html#aurora-serverless-v2.how-it-works.cluster_topology

>  For each of your Aurora DB clusters, you can choose any combination of Aurora Serverless v2 capacity, provisioned capacity, or both. 
>
> You can set up a cluster that contains both Aurora Serverless v2 and provisioned capacity, called a mixed-configuration cluster.

Has this mixed-configuration cluster type been tested?

I see only `ServerlessCluster` construct being touched. But ultimately, I think in V2, the cluster is not serverless, but instances can be serverless. Or am I misunderstanding it?

File: packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts (hunk 2 above)
Comment (pahud) [reply packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts:137]: Agree. Actually that was very similar to my previous design. Let me try again.

Comment (pahud) [reply packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts:137]: Hi @moltar 

Let's make it a little bit clear.

What I am trying to propose is to reuse the `ServerlessCluster` to create either v1 or v2 clusters. As v1 and v2 have different scaling configuration logic and can't be shared, we would have both scaling interfaces for v1 and v2 in the `ServerlessClusterProps` if we have to reuse the `ServerlessCluster` construct for both scenarios. Your sample seems to create `ServerlessClusterV2` just for v2 in addition to the existing `ServerlessCluster` just for v1. I am fine with that as this is opinionated. Is this something you prefer in the API design?

Comment (moltar) [reply packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts:137]: In my example, I am following your proposal to reuse the same class too. I guess my point about that was not very clear.

I was proposing to use [discriminating unions](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) to type narrow the interface based on the `version` prop.

**Comment by moltar** [comment]
> and in v1 adding instances in the cluster is not allowed and not required

Yeah, we've used v1 extensively. I am familiar with this limitation that's why I was asking.

It feels like maybe trying to fit a round peg in a square hole.

- V1 provisioned the cluster _and_ instance, without providing any control over it.
- V2 is essentially going back to the traditional RDS cluster `has_many` instances model.

Meanwhile, we are trying to make ServerlessCluster class represent what it was not intended to do.

Wouldn't it make more sense to make the original cluster work with the v2 instances, and then have some convenience interface, perhaps, over it?

But I might be totally off base here, as I feel like I am still not fully understanding the problem space. So I am sorry, if what I am saying is noise.

**Comment by pahud** [comment]
Hi @moltar please check out this sample in which I created 1 x v1 cluster and 2 x v2 clusters mixed with "serverless" and "provisioned" instances.

image

check out my cdk code used to provision the 3 clusters [here](https://github.com/pahud/aws-cdk/blob/pahud/rds-support-for-aurora-serverless-20197/packages/%40aws-cdk/aws-rds/test/integ.serverlessv2-cluster.ts)

**Comment by pahud** [comment]
Looking at the doc again, I think I was wrong. There is liberally no "Aurora Serververless v2 cluster", only the old aurora cluster. We should be able to add any new "serverless v2 DB instances" or provisioned instances to any new or existing aurora clusters. I will update my PR shortly.

**Comment by moltar** [comment]
I don't think there was even a v1 Serverless Cluster technically. It was always just an `AWS::RDS::DBCluster`, and the `ServerlessClass` construct was a convenience wrapper around it.

I think with V1 the design limitation was that you were not able to add _any_ instances to the cluster (even the first one), as it was created for you. So maybe that's why a separate class was created for it.

**Comment by pahud** [comment]
@moltar thanks for clearing it up. As the `ServerlessCluster` is for serverless v1 only, the DBCluster should satisfy the following scenarios:

1. DBCluster with 1 master and 0 or more replica instances(we already can satisfy this with the `DatabaseCluster` L2)
2. DBCluster with 1 serverless v2 write instance only
3. DBCluster with 1 serverless v2 write and 1 or more serverless v2 read instances
4. DBCluster with 1 serverless v2 write and 1 or more provisioned read instances
5. DBCluster with 1 provisioned write and 1 or more serverless v2 read instances

Maybe we should extend the `DatabaseCluster` class to satisfy all above. With that being said, 
1. `DatabaseCluster` should allow `0` instances(for scenario 3).
2. `DatabaseCluster` should add the `ServerlessV2ScalingConfiguration` support(for scenario 2-5).

wdyt?

**Comment by moltar** [comment]
I don't know all of the permutations of cluster & instances. But I _think_ overall it sounds good to me what you are suggesting.

I think standardizing on the `DBCluster` is the right move too, because v1 will be EOL anyways, and we could probably just put the entire `ServerlessCluster` class into `@deprecated` mode.

**Comment by moltar** [comment]
IMO, the entire `DatabaseCluster` is maybe just designed awkwardly, and does not follow the design principles of CDK.

In ideal world, what I would like to see is (and this is definitely outside of this PR scope):

```
new DatabaseCluster(scope, 'Cluster', {
  // ...

  instances: [
    new ProvisionedInstance(),
    new SeverlessInstance(),
  ]

  // ...
})
```

IMO, we don't need the `instances` prop. Just specify an empty array, and the cluster is launched without instances. And don't need the `instanceProps` prop also, as that seems to be just applying the config to the first instance.

And the convenience method shall not also care about what type of instance is being added. It should just accept an `IInstance` interface, which figures out under the hood.

```ts
cluster.addInstance(new ProvisionedInstance())
// OR
cluster.addInstance(new SeverlessInstance())
```

I have not studied all of the typings, so I am not sure if this is possible. But this feels like the CDK approach.

File: packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts (hunk 2 above)
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-rds/lib/serverless-cluster.ts]: Union types don't work with jsii.

File: packages/@aws-cdk/aws-rds/README.md [hunk 3]
```diff
@@ -673,6 +676,105 @@ new rds.ServerlessClusterFromSnapshot(this, 'Cluster', {
...
+  instanceProps: {
+    vpc,
+    vpcSubnets: {
+      subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
+    },  
+  },
+});
+
+// adding a serverless writer
+cluster.addServerlessInstance('instance1', {
```
Comment (corymhall) [inline packages/@aws-cdk/aws-rds/README.md:714]: Is this adding a serverless v1 or v2 instance? I think we need to update all references to include `V2` if it is referring to `v2` (docs & API), otherwise I am not sure if it is referring to `v1` or `v2`. You might know that I can't add a serverless v1 instance, but others might not.

Is it adding a writer because it is the first instance that we are adding?

File: packages/@aws-cdk/aws-rds/lib/instance-engine.ts [hunk 4]
```diff
@@ -635,6 +636,39 @@ class MySqlInstanceEngine extends InstanceEngineBase {
   }
 }
 
+/**
+ * Properties for Aurora MySQL instance engines.
+ * Used in {@link DatabaseInstanceEngine.auroraMySql}.
+ */
+export interface AuroraMysqlInstanceEngineProps {
+  /** The exact version of the engine to use. */
+  // readonly version: MysqlEngineVersion;
```
Comment (corymhall) [inline packages/@aws-cdk/aws-rds/lib/instance-engine.ts:645]:
```suggestion
```

File: packages/@aws-cdk/aws-rds/README.md [hunk 5]
```diff
@@ -673,6 +676,105 @@ new rds.ServerlessClusterFromSnapshot(this, 'Cluster', {
...
+const cluster = new rds.DatabaseCluster(stack, 'cluster', {
+  engine: rds.DatabaseClusterEngine.auroraMysql({
+    version: rds.AuroraMysqlEngineVersion.VER_3_02_1,
+  }),
+  serverlessV2Scaling: {
+    maxCapacity: 1,
+    minCapacity: 0.5,
+  },
+  // do not create any provisioned instance in the cluster
+  instances: 0,
```
Comment (corymhall) [inline packages/@aws-cdk/aws-rds/README.md:704]: In order to create this cluster with a serverless v2 writer we have to set `instances=0`?

**Review by corymhall — CHANGES_REQUESTED** [review]
I think we may want to think a little bit more about the user experience here. Previously `DatabaseCluster` created _only_ provisioned instances and now it can create a mix of provisioned or serverlessV2. In your proposed API there are too many things that the user needs to keep track of.

- If I start with a provisioned cluster and want to add a serverlessv2 instance can I just do `addServerlessInstance`? or do I also have to go back and add `serverlessV2Scaling` to the cluster props?
- What is `instanceProps` used for if I set `instances=0`? 
- It is not straight forward to figure out if my writer is a serverlessv2 instance or not.
- What happens if I change my writer from a `provisioned` to a `serverlessV2` (or vice versa)

I think instead of starting from the existing `DatabaseCluster` construct, we should start from the ideal customer experience and work backwards from there. If it allows us to continue using the `DatabaseCluster` construct then great, but if we need to create a new construct `DatabaseClusterV2` then that is ok too.

**Comment by pahud** [comment]
> I think we may want to think a little bit more about the user experience here. Previously `DatabaseCluster` created _only_ provisioned instances and now it can create a mix of provisioned or serverlessV2. In your proposed API there are too many things that the user needs to keep track of.
> 
> * If I start with a provisioned cluster and want to add a serverlessv2 instance can I just do `addServerlessInstance`? or do I also have to go back and add `serverlessV2Scaling` to the cluster props?
> * What is `instanceProps` used for if I set `instances=0`?
> * It is not straight forward to figure out if my writer is a serverlessv2 instance or not.
> * What happens if I change my writer from a `provisioned` to a `serverlessV2` (or vice versa)
> 
> I think instead of starting from the existing `DatabaseCluster` construct, we should start from the ideal customer experience and work backwards from there. If it allows us to continue using the `DatabaseCluster` construct then great, but if we need to create a new construct `DatabaseClusterV2` then that is ok too.

Yes I was struggling if we should use the existing `DatabaseCluster` for both common cluster and v2-enabled cluster or if we really need to create a new `DatabaseClusterV2` for serverless v2 support.

Current API design is based on using the existing `DatabaseCluster` construct without potential breaking changes. In fact there is no so-called `serverless v2 cluster` but just `serverless v2-enabled` cluster. If we need to create a cluster with serverless writer and serverless reader(s), the first step is to create a `DatabaseCluster` with `0` instances and then add serverless instances into this cluster. Yes it's a little bit awkward but we do see similar design in `aws-eks` to create a cluster with 0 instances and then [cluster.addNodegroupCapacity()](https://github.com/aws/aws-cdk/blob/68fdfc131b0810eba44659db1272e930f38168cf/packages/%40aws-cdk/aws-eks/lib/cluster.ts#L1632) or [addAutoScalingGroupCapacity()](https://github.com/aws/aws-cdk/blob/68fdfc131b0810eba44659db1272e930f38168cf/packages/%40aws-cdk/aws-eks/lib/cluster.ts#L1590) as EKS Cluster construct allows us to create clusters with defined or empty capacity.  So I think `DatabaseCluster` might have similar design like that.

The user experience could be like:

1. Users are allowed to create a general cluster or `serverlessv2-enabled` cluster with the `DatabaseCluster` construct.
2. For general clusters with writer and read replica(s), users can leave the `instances` property undefined(default is 2) or any other values supported for this scenario. We already can satisfy this with `DatabaseCluster`.
3. For clusters with `serverlessV2-enabled` support, however, we have many possible and supported scenarios like:
   a. 1 serverless writer + 1 or more serverless reader(s)
   b. 1 serverless writer + 1 or more provisioned reader(s)
   c. 1 provisioned writer + 1 or more serverless reader(s)
   d. 1 provisioned writer + 1 or more provision reader(s)
   e. 1 provisioned writer + 1 or more provision reader(s) + 1 or more serverless reader(s)
   f. 1 serverless writer + 1 or more provision reader(s) + 1 or more serverless reader(s)

   To use the existing `DatabaseCluster` for case `a`, `b` and `f` we have to set `instances` to `0` as the first instance would be the writer and we can't create the serverless writer with current `DatabaseCluster` construct and that's where `addServerlessInstance()` comes in for those scenarios.
   
   For case `c` we should create the `DatabaseCluster` with `instances: 1` so the first instance would be the provisioned writer instance and `addServerlessInstance()` to create the serverless reader.
   
   For case `d` we just specify `instances: 2` for the `serverlessV2-enabled` cluster.
 
  For case `e` we specify `instances` >=2 and then `addServerInstance()` to create additional reader(s).
   
To answer your questions above:
> * If I start with a provisioned cluster and want to add a serverlessv2 instance can I just do `addServerlessInstance`? or do I also have to go back and add `serverlessV2Scaling` to the cluster props?

If you create a cluster without `serverlessv2 enabled`, you **can't** `addServerlessInstance()` to it. We can check the eligibility in this method and throw to avoid that. 

> * What is `instanceProps` used for if I set `instances=0`?

`instanceProps` will be used when we `addServerlessInstance` or `addInstance()` for the `serverlessV2-enabled` cluster to reference the `vpc` property from it.

> * It is not straight forward to figure out if my writer is a serverlessv2 instance or not.

For serverlessv2 writer(discussed in case a. b. and f. above) we have to set `instances: 0` and explicitly `addServerlessInstance()` from the empty cluster to create our first instance as the serverless writer.

> * What happens if I change my writer from a `provisioned` to a `serverlessV2` (or vice versa)

This is interesting. From CFN's perspective I believe this will trigger the update replacement described [in this doc](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbinstance.html). CFN will create a new one before it "replaces" the old one, however, terminating the writer for replacement should result in a "failover" so your writer might immediately failover to the existing reader. I am not sure how CFN deals with that.  Will give it a try and update here.

I hope this helps to make it a little bit clear. I agree we should consider the user experience and work backwards from it. I am totally fine with a new `DatabaseClusterV2` construct just for `serverlessV2-enabled` clusters and provide `addServerlessInstance` and `addInstance` for it. What do you think?

File: packages/@aws-cdk/aws-rds/README.md (hunk 5 above)
Comment (pahud) [reply packages/@aws-cdk/aws-rds/README.md:704]: Yes 😄

**Comment by pahud** [comment]
Hi @corymhall 

Last night I implemented a new `ServerlessV2DatabaseCluster` as a different API design for this PR which makes everything more clean and easy to understand. While `DatabaseCluster` is untouched in this design, the new `ServerlessV2DatabaseCluster` construct looks like:

```ts
/**
 * Aurora serverless v2 cluster for MySQL with the mix of serverless and provisioned instances.
 */
// Aurora Serverless v2 cluster for MySQL
new rds.ServerlessV2DatabaseCluster(stack, 'cluster2', {
  engine: rds.DatabaseClusterEngine.auroraMysql({ version: rds.AuroraMysqlEngineVersion.VER_3_02_1 }),
  serverlessV2Scaling: {
    maxCapacity: 4,
    minCapacity: 0.5,
  },
  // 1 writer(serverless)
  writer: { engine: rds.DatabaseInstanceEngine.auroraMysql({ version: rds.AuroraMysqlEngineVersion.VER_3_02_1 }) },
  // 2 readers: 1 serverless and 1 provisioned.
  reader: [
    { engine: rds.DatabaseInstanceEngine.auroraMysql({ version: rds.AuroraMysqlEngineVersion.VER_3_02_1 }) },
    {
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.R5, ec2.InstanceSize.LARGE),
      engine: rds.DatabaseInstanceEngine.auroraMysql({ version: rds.AuroraMysqlEngineVersion.VER_3_02_1 }),
    },
  ],
  vpc,
  vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS },
  removalPolicy: cdk.RemovalPolicy.DESTROY,
});
```
Should we opt in a new construct like this for this PR?

**Comment by moltar** [comment]
IMO, the solution I have proposed is more idiomatic:

```ts
new DatabaseCluster(scope, 'Cluster', {
  // ...

  instances: [
    new ProvisionedInstance(),
    new SeverlessInstance(),
  ]

  // ...
})
```

If you want to have a specific reader writer instance, you can create classes for that:

```ts
new DatabaseCluster(scope, 'Cluster', {
  // ...

  instances: [
    new ReaderInstance(),
    new WriterInstance(),
  ]

  // ...
})
```

**Comment by corymhall** [comment]
> If you create a cluster without serverlessv2 enabled, you can't addServerlessInstance() to it. We can check the eligibility in this method and throw to avoid that.

If this is the case then I definitely think that we should go with a new construct. This means that there are 3 different types of clusters that you can create, ServerlessV1 cluster, traditional provisioned cluster, serverless v2 enabled cluster. But based on what I can see in the docs it looks like it _is_ possible to add serverless v2 instances to an existing cluster.

> Last night I implemented a new ServerlessV2DatabaseCluster as a different API design for this PR which makes everything more clean and easy to understand. While DatabaseCluster is untouched in this design, the new ServerlessV2DatabaseCluster construct looks like:

I think instead of `ServerlessV2DatabaseCluster` we should call it something like `DatabaseClusterV2`. Even though users _can_ still create a provisioned cluster with `DatabaseCluster` it doesn't sound like there is a reason why you would want to instead of creating a serverlessV2 enabled cluster.

I have some outstanding questions that I think we need to answer before we can design the experience.

> For clusters with serverlessV2-enabled support, however, we have many possible and supported scenarios like:

1. These are _possible_ scenarios, but I am curious what the actual use cases are. _Why_ would I want to created a mixed use cluster? What are the benefits/drawbacks.

From the [docs](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.html)
> The granularity of scaling in Aurora Serverless v2 helps you to match capacity closely to your database's needs. For a provisioned cluster, scaling up requires adding a whole new DB instance. For an Aurora Serverless v1 cluster, scaling up requires doubling the number of Aurora capacity units (ACUs) for the cluster, such as from 16 to 32 or 32 to 64. In contrast, Aurora Serverless v2 can add half an ACU when only a little more capacity is needed. It can add 0.5, 1, 1.5, 2, or additional half-ACUs based on the additional capacity needed to handle an increase in workload. And it can remove 0.5, 1, 1.5, 2, or additional half-ACUs when the workload decreases and that capacity is no longer needed. 

Do I even need to create more than 1 serverless v2 instance (I think the answer is yes)?

> You can modify existing DB instances from provisioned to Aurora Serverless v2 or from Aurora Serverless v2 to provisioned. You don't need to create a new cluster or a new DB instance in such cases. 

I wonder how this works?

> Suppose that you already have an Aurora application running on a provisioned cluster. You can check how the application would work with Aurora Serverless v2 by adding one or more Aurora Serverless v2 DB instances to the existing cluster as reader DB instances. You can check how often the reader DB instances scale up and down. You can use the Aurora failover mechanism to promote an Aurora Serverless v2 DB instance to be the writer and check how it handles the read/write workload. That way, you can switch over with minimal downtime and without changing the endpoint that your client applications use. For details on the procedure to convert existing clusters to Aurora Serverless v2, see 

It looks like you can convert an existing cluster to serverless v2?

2.  If there are valid mixed use scenarios, then how do I account for things like failover? If there is a valid reason why I would want a serverlessv2 writer and a provisioned reader, what happens when I failover and the provisioned instance becomes my writer? Does it create a new serverlessv2 reader, or does it create a new provisioned reader?

Some of this is covered in [scaling](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.how-it-works.html#aurora-serverless-v2.how-it-works.scaling) (i.e. promotion tiers).

**Comment by mrgrain** [comment]
@corymhall
>These are possible scenarios, but I am curious what the actual use cases are. Why would I want to created a mixed use cluster? What are the benefits/drawbacks.
One typical reason will be cost. Aurora Serverless V2 is significantly more expensive than Provisioned On-Demand instances. In many scenarios a provisioned (and reserved) base capacity will be more economical.

**Comment by pahud** [comment]
@corymhall  Love the discussion! Appreciated! Let's dive a little bit deeper and get things sorted.

> > If you create a cluster without serverlessv2 enabled, you can't addServerlessInstance() to it. We can check the eligibility in this method and throw to avoid that.
> 
> If this is the case then I definitely think that we should go with a new construct. This means that there are 3 different types of clusters that you can create, ServerlessV1 cluster, traditional provisioned cluster, serverless v2 enabled cluster. But based on what I can see in the docs it looks like it _is_ possible to add serverless v2 instances to an existing cluster.

From CFN's perspective, there's only one [AWS::RDS::DBCluster](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbcluster.html) resource and we can create 3 types of clusters from it:

1. **general clusters with a writer and replica(s)** - instance-based clusters w/o serverlessV1 or serverlessV2-enabled. From CFN's perspective, both `ServerlessV2ScalingConfiguration` and `ScalingConfiguration` should not be specified.
2. **serverless v1 clusters** - no instances at all. Provisioned with `ScalingConfiguration` specified.
3. **serverless v2 clusters or clusters with serverless v2 support** - mix with serverless and/or provisioned instance(s) writer and reader. Provisioned with `ServerlessV2ScalingConfiguration` specified.

From my experiment with CDK, you can't just add a `serverless v2 instance` into a general cluster or v1-enabled cluster. You will receive an error message like `ServerlessV2ScalingConfiguration` is not enabled in the existing cluster. I don't think CFN will allow you in-place upgrade a general cluster to v2 or v1 to v2 by simply adding the `ServerlessV2ScalingConfiguration` optional prop in the construct because according to the [doc](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.upgrade.html#aurora-serverless-v2.move-from-serverless-v1), moving/upgrading from a [provisioned cluster to v2](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.upgrade.html#aurora-serverless-v2.switch-from-provisioned) or [v1 to v2](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.upgrade.html#aurora-serverless-v2.move-from-serverless-v1) has some limitations and conditions. I will call out this in the aws-rds README in this PR.

> 
> I have some outstanding questions that I think we need to answer before we can design the experience.
> 
> 1. These are _possible_ scenarios, but I am curious what the actual use cases are. _Why_ would I want to created a mixed use cluster? What are the benefits/drawbacks.
> 
I'm probably not the best one to answer this question but AFAIK 
1. the provisioned one might be more economical compared to the serverless one. If customer use case is a predictable steady read workload and does not require autoscaling with unpredictable write/update workload, a provisioned reader with a serverless writer might be a better option than the serverless writer+reader.
2. Similarly, predictable intensive write/update with unpredictable read scenarios,  provisioned writer + serverless reader may be a good choice.
> 
> Do I even need to create more than 1 serverless v2 instance (I think the answer is yes)?
As aurora serverless v2 is single-AZ based instance. Customers may opt in multi-AZ serverless readers, in which case they might need 2 or more serverless readers across multi-AZ. One of the advantages is the AZ-tolerance, another one is that when the primary serverless reader instance fails(everything fails 😄 ), it could immediate fail over to another serverless reader, instead of the writer.

> 
> > You can modify existing DB instances from provisioned to Aurora Serverless v2 or from Aurora Serverless v2 to provisioned. You don't need to create a new cluster or a new DB instance in such cases.
> 
> I wonder how this works?
> 
I probably need to look into that as it seems not possible from console. 

> > Suppose that you already have an Aurora application running on a provisioned cluster. You can check how the application would work with Aurora Serverless v2 by adding one or more Aurora Serverless v2 DB instances to the existing cluster as reader DB instances. You can check how often the reader DB instances scale up and down. You can use the Aurora failover mechanism to promote an Aurora Serverless v2 DB instance to be the writer and check how it handles the read/write workload. That way, you can switch over with minimal downtime and without changing the endpoint that your client applications use. For details on the procedure to convert existing clusters to Aurora Serverless v2, see
> 
> It looks like you can convert an existing cluster to serverless v2?
yes, as discussed above, the `provisioned cluster`, `serverless v1 cluster` and the `serverless v2 cluster` are essentially the same `AWS::RDS::DBCluster` with different mutual-exclusive props.

> 
> 2. If there are valid mixed use scenarios, then how do I account for things like failover? If there is a valid reason why I would want a serverlessv2 writer and a provisioned reader, what happens when I failover and the provisioned instance becomes my writer? Does it create a new serverlessv2 reader, or does it create a new provisioned reader?

With a serverlessV2 writer and one provisioned reader:
1. You have a single-AZ risk for both writer and reader. If the reader fails, I believe the read traffic will switch to the writer and a new reader might be start provisioning and switch back the read traffic when the new reader is ready.
2. if the serverless writer fails I believe the provisioned reader will be promoted as the provisioned writer with a little period of writing interruptions. I am not sure if the control plane will create a new serverless writer instance and switch back the write traffic.

However, customers with very critical, steady and intensive read queries might consider multiple provisioned readers across multi-AZ for fault tolerance, which will not impact the writer when any single reader fails.

Nice discussion. Thanks for bring it up.
> 
> Some of this is covered in [scaling](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.how-it-works.html#aurora-serverless-v2.how-it-works.scaling) (i.e. promotion tiers).

**Comment by corymhall** [comment]
> > These are possible scenarios, but I am curious what the actual use cases are. Why would I want to created a mixed use cluster? What are the benefits/drawbacks.
> 
> One typical reason will be cost. Aurora Serverless V2 is significantly more expensive than Provisioned On-Demand instances. In many scenarios a provisioned (and reserved) base capacity will be more economical.

Yep! I think my point in bringing this up was that I think our design should take these specific scenarios into consideration. For example, If I _really_ want to (or even have to) have a provisioned instance as my writer, then I also need a provisioned instance as a reader in the 0 promotion tier and any serverlessv2 instances should be in lower promotion tiers. If we make this hard for the user to do then we could make it easy for users to shoot themselves in the foot (e.g. Only a provisioned writer can handle the production load and they failover to a serverless v2 instance and prod goes down :( ). Previously you could only create a cluster with a single type of instance so these types of considerations didn't really matter.

> From my experiment with CDK, you can't just add a serverless v2 instance into a general cluster or v1-enabled cluster. You will receive an error message like ServerlessV2ScalingConfiguration is not enabled in the existing cluster. I don't think CFN will allow you in-place upgrade a general cluster to v2 or v1 to v2 by simply adding the ServerlessV2ScalingConfiguration optional prop in the construct because according to the [doc](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.upgrade.html#aurora-serverless-v2.move-from-serverless-v1), moving/upgrading from a [provisioned cluster to v2](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.upgrade.html#aurora-serverless-v2.switch-from-provisioned) or [v1 to v2](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.upgrade.html#aurora-serverless-v2.move-from-serverless-v1) has some limitations and conditions to consider. I will call out this in the aws-rds README in this PR.

My point in bringing this up is that we should take the upgrade experience into consideration. If it **is** possible to upgrade an existing cluster, then maybe we don't want to require the user to use a new construct since that would prevent them from upgrading their cluster. If it is **not** possible then they have to create a new cluster anyway.

**Review by corymhall — CHANGES_REQUESTED** [review]
I think this feature might be a good candidate for an RFC.

**Comment by pahud** [comment]
Some important findings today:

1. You actually _can_ add `serverless v2 instances` into an existing cluster provisioned by the existing `rds.DatabaseCluster` construct w/o the `ServerlessV2ScalingConfiguration` prop of the cluster.

```ts
// create a ordinary provisioned cluster
new rds.DatabaseCluster(stack, 'general-cluster', {
  // make sure you use serverless v2 compatible DatabaseClusterEngine
  engine: rds.DatabaseClusterEngine.auroraMysql({ version: rds.AuroraMysqlEngineVersion.VER_3_02_1 }),
  instanceProps: {
    vpc,
    vpcSubnets,
  },
  subnetGroup,
});

// OK now you are allowed to add a serverless v2 reader into this cluster from console.
```
This indicates that you are not literally _convert_ the original cluster into a `serverless v2 cluster`. They are exactly the same cluster.  To add a serverless v2 reader into an existing cluster, just make sure the existing cluster is configured with v2-compatible engine(mysql 8, postgres 13 and 14), which is a known condition described in the doc.

>

In this case, `ServerlessV2ScalingConfiguration` is not strictly required for cluster to add serverless v2 instances. And users previously created their provisioned cluster with the existing `rds.DatabaseCluster` construct with v2-compatible cluster engine should be allowed to add serverless v2 instances in their existing clusters.

2. In the DBCluster with v2 compatible cluster engine, you are allowed to in-place "convert" the existing instances from provisioned writer/reader to serverless writer/reader or vice versa.  On instance modification, the instance enters the "Modifying" status in the console. I am not sure if the traffic will be switched off but obviously it can be in-place converted between provisioned and serverless instances. Amazing!

If we consider the existing clusters created with the `rds.DatabaseCluster` construct, we should allow them to:

1. add a `serverless v2 reader` in addition to the existing provisioned writer and reader with CDK in the existing cluster.
2. convert the existing provisioned writer/reader to serverless v2.
3. enable the `ServerlessV2ScalingConfiguration` for the cluster.

Apart from above, we should allow them to create a new cluster with both serverless v2 writer and reader(s).

Sounds like we should extend the existing DatabaseCluster construct instead?

**Comment by corymhall** [comment]
I'm going to go ahead and close this for now since we are going the RFC route.

**Comment by ThuF** [comment]
@corymhall, @pahud, What's happening with the support for Aurora Serverless v2 construct?
